File: api/app/application/services/session_service.py
from app.domain.external.sandbox import Sandbox
from app.interface.schemas.event import EventMapper
from app.interface.schemas.session import GetSessionResponse
from app.domain.repositories.uow import IUnitOfWork
from typing import Callable
from app.application.errors.exceptions import NotFoundException
from app.domain.models.session import Session


class SessionService:
    def __init__(
        self, uow_factory: Callable[[], IUnitOfWork], sandbox_cls: Callable[[], Sandbox]
    ):
        self.uow_factory = uow_factory
        self.uow: IUnitOfWork = uow_factory()
        self.sandbox_cls: Callable[[], Sandbox] = sandbox_cls

    async def create_session(self):
        session = Session(title="新对话")

        async with self.uow:
            await self.uow.session.save(session)
        return session

    async def get_session(self, session_id: str):
        async with self.uow:
            session = await self.uow.session.get_by_id(session_id)

        if not session:
            raise NotFoundException(msg=f"找不到此会话{session_id}")

        return session

    async def get_sessions(self):
        async with self.uow:
            return await self.uow.session.get_all()

    async def clear_unread_msg_count(self, session_id: str):
        async with self.uow:
            await self.uow.session.update_unread_msg_count(session_id, 0)

    async def delete_session(self, session_id: str):
        async with self.uow:
            session = await self.uow.session.get_by_id(session_id)

            if not session:
                raise NotFoundException(msg=f"找不到此会话{session_id}")

            await self.uow.session.delete(session_id)

    async def update_session_title(self, session_id: str, title: str):
        async with self.uow:
            session = await self.uow.session.get_by_id(session_id)

            if not session:
                raise NotFoundException(msg=f"找不到此会话{session_id}")

            await self.uow.session.update_title(session_id, title)

    async def get_session_files(self, session_id: str):
        async with self.uow:
            session = await self.uow.session.get_by_id(session_id)

            if not session:
                raise NotFoundException(msg=f"找不到此会话{session_id}")

            return session.files

    async def get_session_file(self, session_id: str, filepath: str):
        async with self.uow:
            session = await self.uow.session.get_by_id(session_id)

            if not session:
                raise NotFoundException(msg=f"找不到此会话{session_id}")

            if not session.sandbox_id:
                raise NotFoundException(msg=f"此会话未创建沙箱")
            sandbox: Sandbox = await self.sandbox_cls.get(session.sandbox_id)
            if not sandbox:
                raise NotFoundException(msg=f"找不到或者销毁了此会话{session_id}的沙箱")
            result = await sandbox.file_read(filepath)
            if not result.success:
                raise NotFoundException(msg=f"读取文件失败:{result.error}")
            return result.data

    async def read_shell_output(self, session_id: str, shell_session_id: str):
        async with self.uow:
            session = await self.uow.session.get_by_id(session_id)

            if not session:
                raise NotFoundException(msg=f"找不到此会话{session_id}")

            if not session.sandbox_id:
                raise NotFoundException(msg=f"此会话未创建沙箱")
            sandbox: Sandbox = await self.sandbox_cls.get(session.sandbox_id)
            if not sandbox:
                raise NotFoundException(msg=f"找不到或者销毁了此会话{session_id}的沙箱")
            result = await sandbox.view_shell(shell_session_id, True)
            if not result.success:
                raise NotFoundException(msg=f"读取shell失败:{result.error}")
            return result.data

    async def get_sandbox_vnc_url(self, session_id: str):
        async with self.uow:
            session = await self.uow.session.get_by_id(session_id)

            if not session:
                raise NotFoundException(msg=f"找不到此会话{session_id}")

            if not session.sandbox_id:
                raise NotFoundException(msg=f"此会话未创建沙箱")
            sandbox: Sandbox = await self.sandbox_cls.get(session.sandbox_id)
            if not sandbox:
                raise NotFoundException(msg=f"找不到或者销毁了此会话{session_id}的沙箱")
            return sandbox.vnc_url

File: api-node/src/services/session.service.ts
import { v4 as uuidv4 } from 'uuid';
import { NotFoundException } from '../errors/AppException';
import { FileRecord, Session, SessionStatus } from '../domain/models';
import { DockerSandbox } from '../infrastructure/sandbox/docker-sandbox';
import { sessionRepository, SessionRepository } from '../repositories/session.repository';

export class SessionService {
  constructor(private readonly repo: SessionRepository = sessionRepository) {}

  async createSession(): Promise<Session> {
    const session: Session = {
      id: uuidv4(),
      title: '新对话',
      unread_msg_count: 0,
      latest_msg: '',
      latest_message_at: new Date(),
      events: [],
      files: [],
      memories: {},
      status: SessionStatus.PENDING,
      created_at: new Date(),
      updated_at: new Date(),
    };
    await this.repo.save(session);
    return session;
  }

  async getSession(sessionId: string): Promise<Session> {
    const session = await this.repo.getById(sessionId);
    if (!session) throw new NotFoundException(`找不到此会话${sessionId}`);
    return session;
  }

  async getSessions(): Promise<Session[]> {
    return this.repo.getAll();
  }

  async clearUnreadMsgCount(sessionId: string): Promise<void> {
    await this.repo.updateUnreadMsgCount(sessionId, 0);
  }

  async deleteSession(sessionId: string): Promise<void> {
    await this.getSession(sessionId);
    await this.repo.delete(sessionId);
  }

  async updateSessionTitle(sessionId: string, title: string): Promise<void> {
    await this.getSession(sessionId);
    await this.repo.updateTitle(sessionId, title);
  }

  async getSessionFiles(sessionId: string): Promise<FileRecord[]> {
    const session = await this.getSession(sessionId);
    return session.files;
  }

  async getSessionFile(sessionId: string, filepath: string) {
    const session = await this.getSession(sessionId);
    if (!session.sandbox_id) throw new NotFoundException('此会话未创建沙箱');
    const sandbox = await DockerSandbox.get(session.sandbox_id);
    if (!sandbox) throw new NotFoundException(`找不到或者销毁了此会话${sessionId}的沙箱`);
    const result = await sandbox.fileRead(filepath);
    if (!result.success) throw new NotFoundException(`读取文件失败:${result.message}`);
    return result.data;
  }

  async readShellOutput(sessionId: string, shellSessionId: string) {
    const session = await this.getSession(sessionId);
    if (!session.sandbox_id) throw new NotFoundException('此会话未创建沙箱');
    const sandbox = await DockerSandbox.get(session.sandbox_id);
    if (!sandbox) throw new NotFoundException(`找不到或者销毁了此会话${sessionId}的沙箱`);
    const result = await sandbox.viewShell(shellSessionId, true);
    if (!result.success) throw new NotFoundException(`读取shell失败:${result.message}`);
    return result.data;
  }

  async getSandboxVncUrl(sessionId: string): Promise<string> {
    const session = await this.getSession(sessionId);
    if (!session.sandbox_id) throw new NotFoundException('此会话未创建沙箱');
    const sandbox = await DockerSandbox.get(session.sandbox_id);
    if (!sandbox) throw new NotFoundException(`找不到或者销毁了此会话${sessionId}的沙箱`);
    return sandbox.vncUrl;
  }
}

export const sessionService = new SessionService();
